kani-driver/src/cbmc_output_parser.rs

```diff
@@ -0,0 +1,989 @@
+fn filepath(file: String) -> String {
+    let file_path = PathBuf::from(file.clone());
+    let cur_dir = env::current_dir().unwrap();
+
+    let diff_path_opt = diff_paths(file_path, cur_dir);
+    if let Some(diff_path) = diff_path_opt {
+        diff_path.into_os_string().into_string().unwrap()
+    } else {
+        file
+    }
+}
+/// Struct that represents traces.
+///
+/// In general, traces may include more information than this, but this is not
+/// documented anywhere. So we ignore the rest for now.
+#[derive(Clone, Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct TraceItem {
+    pub thread: u32,
+    pub step_type: String,
+    pub hidden: bool,
+    pub source_location: Option<SourceLocation>,
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "UPPERCASE")]
+pub enum CheckStatus {
+    Failure,
+    Success,
+    Undetermined,
+    Unreachable,
+}
+
+impl std::fmt::Display for CheckStatus {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let check_str = match self {
+            CheckStatus::Success => style("SUCCESS").green(),
+            CheckStatus::Failure => style("FAILURE").red(),
+            CheckStatus::Unreachable => style("UNREACHABLE").yellow(),
+            CheckStatus::Undetermined => style("UNDETERMINED").yellow(),
+        };
+        write! {f, "{}", check_str}
+    }
+}
+
+#[derive(PartialEq)]
+enum Action {
+    ClearInput,
+    ProcessItem,
+}
+
+/// A parser for CBMC output, whose state is determined by:
+///  1. The input accumulator, required to process items on the fly.
+///  2. The buffer, which is accessed to retrieve more lines.
+///
+/// CBMC's JSON output is defined as a JSON array which contains:
+///  1. One program at the beginning (i.e., a message with CBMC's version).
+///  2. Messages, which can appears anywhere, and are either status messages or error messages.
+///  3. Verification results, another JSON array with all individual checks.
+///  4. Prover status, at the end. Because the verification results depends on
+///     our postprocessing, this is not used.
+///
+/// The parser reads the output line by line. A line may trigger one action, and
+/// the action may return a parsed item.
+struct Parser<'a, 'b> {
+    pub input_so_far: String,
+    pub buffer: &'a mut BufReader<&'b mut ChildStdout>,
+}
+
+impl<'a, 'b> Parser<'a, 'b> {
+    pub fn new(buffer: &'a mut BufReader<&'b mut ChildStdout>) -> Self {
+        Parser { input_so_far: String::new(), buffer }
+    }
+
+    /// Triggers an action based on the input:
+    ///  * Square brackets ('[' and ']') will trigger the `ClearInput` action
+    ///    because we assume parsing is done on a JSON array.
+    ///  * Curly closing bracket ('}') preceded by two spaces will trigger the
+    ///    `ProcessItem` action.
+    fn triggers_action(&self, input: String) -> Option<Action> {
+        if input.starts_with('[') || input.starts_with(']') {
```

Comment:
This may ignore any characters that appear after `[` or `]`. If CBMC's current output never includes characters, perhaps we should assert that the line only has `[` or `]`?